src/index.ml

```diff
@@ -688,7 +688,10 @@ struct
     flush_instance ~no_async:() ~with_fsync:true t;
 
     let go () =
-      hook `Before;
+      (try hook `Before
+       with exn ->
+         Semaphore.release t.merge_lock;
+         raise exn);
```

Comment:
maybe define `hook` to always release the hook? That'll avoid code duplication.

Reply:
I'll wrap the entire thing in a release to solve https://github.com/mirage/index/issues/284 at the same time.